app/gui/src/controller/searcher/breadcrumbs.rs

```diff
@@ -26,18 +30,145 @@ impl Breadcrumbs {
+    /// is selected.
+    pub fn selected(&self) -> Option<component::Id> {
+        if self.is_top_module() {
+            None
+        } else {
+            let index = self.selected.get();
+            self.list.borrow().get(index - 1).map(BreadcrumbEntry::id)
+        }
+    }
+}
+
+
+
+// =======================
+// === BreadcrumbEntry ===
+// =======================
+
+/// A single entry in the breadcrumbs panel.
+#[derive(Debug, Clone)]
+pub struct BreadcrumbEntry {
+    displayed_name: ImString,
+    component_id:   component::Id,
+    qualified_name: QualifiedName,
+}
+
+impl BreadcrumbEntry {
+    /// A displayed label of the entry.
+    pub fn name(&self) -> ImString {
+        self.displayed_name.clone_ref()
+    }
+
+    /// A component id of the entry.
+    pub fn id(&self) -> component::Id {
+        self.component_id
+    }
+
+    /// A qualified name of the entry.
+    pub fn qualified_name(&self) -> &QualifiedName {
+        &self.qualified_name
+    }
+}
+
+impl From<(component::Id, Rc<Entry>)> for BreadcrumbEntry {
+    fn from((component_id, entry): (component::Id, Rc<Entry>)) -> Self {
+        let qualified_name = entry.qualified_name();
+        let displayed_name = ImString::new(&entry.name);
+        BreadcrumbEntry { displayed_name, component_id, qualified_name }
+    }
+}
+
+
+
+// ===============
+// === Builder ===
+// ===============
+
+/// A builder for the breadcrumbs list. It is used to include all parent modules when pushing the
+/// new breadcrumb to the panel.
+#[derive(Debug)]
+pub struct Builder<'a> {
+    database:   &'a model::SuggestionDatabase,
+    components: component::List,
+}
+
+impl<'a> Builder<'a> {
+    /// Constructor.
+    pub fn new(database: &'a model::SuggestionDatabase, components: component::List) -> Self {
+        Self { database, components }
+    }
+
+    /// Build a list of breadcrumbs for a specified module. The list will contain:
+    /// 1. The main module of the project.
+    /// 2. All parent modules of the [`module`].
+    /// 3. The [`module`] itself.
+    ///
+    /// Returns [`None`] if the [`module`] is not found in the database or in the components list.
+    pub fn build(self, module: &component::Id) -> Option<Vec<BreadcrumbEntry>> {
+        let mut result = Vec::new();
+        let module_name = self.components.module_qualified_name(*module)?;
+        let entry = BreadcrumbEntry::from(self.lookup(&module_name)?);
+        result.push(entry);
+        result.extend(self.collect_parents(&module_name));
+        let project_name = module_name.project_name.clone();
+        let main_module = module::QualifiedName::new_main(project_name.clone());
+        if let Some(entry) = self.lookup(&main_module) {
+            result.push(BreadcrumbEntry {
+                displayed_name: String::from(project_name.project).into(),
+                ..entry.into()
+            });
+        }
+        Some(result.reversed())
```

Comment:
I strongly opt for avoid mutable vars where possible. In this case we can chain iterators `once(entry)` with result of (`sellf.collect_parentr`) with main module entry (an option is also an iterator). Then, collect the vector from the chain iterator.